xray/src/client.rs

```diff
@@ -33,34 +33,39 @@ impl Client {
 
     pub async fn do_handshake(&mut self, wg_addr: SocketAddrV4) -> XRayResult<()> {
         println!("Handshake: starting");
-        let tunn = self
-            .tunn
-            .as_mut()
-            .expect("This function should only be called on clients with a Tunn object");
-        match tunn.format_handshake_initiation(&mut self.buf, true) {
-            TunnResult::WriteToNetwork(packet) => {
-                self.sock.send_to(packet, wg_addr).await?;
-            }
-            unexpected => {
-                return Err(XRayError::UnexpectedTunnResult(format!("{unexpected:?}")));
+        'new_handshake: for _ in 0..3 {
+            let tunn = self
+                .tunn
+                .as_mut()
+                .expect("This function should only be called on clients with a Tunn object");
+            match tunn.format_handshake_initiation(&mut self.buf, true) {
+                TunnResult::WriteToNetwork(packet) => {
+                    self.sock.send_to(packet, wg_addr).await?;
+                }
+                unexpected => {
+                    return Err(XRayError::UnexpectedTunnResult(format!("{unexpected:?}")));
+                }
             }
-        }
-        let mut handshake_buf = vec![0; 512];
-        let handshake_timeout = tokio::time::sleep(tokio::time::Duration::from_secs(3));
-        tokio::pin!(handshake_timeout);
-        loop {
-            tokio::select! {
-                Ok(recv_type) = self.recv_encrypted(&mut handshake_buf) => {
-                    if matches!(recv_type, RecvType::HandshakeResponse) {
-                        println!("Handshake: done");
-                        return Ok(());
+            let mut handshake_buf = vec![0; 512];
+            let handshake_timeout = tokio::time::sleep(tokio::time::Duration::from_secs(2));
+            tokio::pin!(handshake_timeout);
+            loop {
+                tokio::select! {
+                    Ok(recv_type) = self.recv_encrypted(&mut handshake_buf) => {
+                        if matches!(recv_type, RecvType::HandshakeResponse) {
+                            println!("Handshake: done");
+                            return Ok(());
+                        }
+                    }
+                    _ = &mut handshake_timeout => {
+                        println!("Handshake timed out trying again...");
+                        continue 'new_handshake;
```

Comment:
I know this was already here, but - can this be simplified and use https://docs.rs/tokio/latest/tokio/time/fn.timeout.html ?

Reply:
Fixed